kedro/io/core.py

```diff
@@ -281,11 +361,49 @@ def save(self: Self, data: _DI) -> None:
         return save
 
     def __init_subclass__(cls, **kwargs: Any) -> None:
-        """Decorate the `load` and `save` methods provided by the class.
+        """Customizes the behavior of subclasses of AbstractDataset during
+        their creation. This method is automatically invoked when a subclass
+        of AbstractDataset is defined.
+
 
+        Decorates the `load` and `save` methods provided by the class.
         If `_load` or `_save` are defined, alias them as a prerequisite.
 
         """
+
+        # Save the original __init__ method of the subclass
+        init_func: Callable = cls.__init__
+
+        def init_decorator(previous_init: Callable) -> Callable:
```

Comment:
I believe you can drop the padding of an extra function, this decorator will not be used in the code as a decorator, but rather you will just assign the function it returns to be the `cls.__init__`, so only the `new_init` function is required.

Reply:
I see what you mean but this is not the correct syntax for `@wraps` cause you need to wrap a function that is passed to the outer function.

```python
init_func: Callable = cls.__init__

def __init_subclass__(cls, **kwargs):
    init_func: Callable = cls.__init__

    def new_init(self, *args, **kwargs) -> None:
        @wraps(init_func)  #  <--- This is not the correct syntax
        if type(self) is cls:
            call_args = getcallargs(init_func, self, *args, **kwargs)
            self.__post_init__(call_args)

    cls.__init__ = new_init
```

But you don't really need to wrap it as well, so I ended up with, since we can simply call `init_func`

```python
init_func: Callable = cls.__init__

def new_init(self, *args, **kwargs) -> None:
    # Call the original __init__ method
    init_func(self, *args, **kwargs)
    # Capture and save the arguments passed to the original __init__
    call_args = getcallargs(init_func, self, *args, **kwargs)
    self._init_args = call_args

cls.__init__ = new_init
```